BlackBoxPredictor OSS: open-source NQL and custom transforms

salexspb opened the pull request:
Summary:
Overal context: open-source BlackBoxPredictor as the entry
point for inference in Caffe2 (thread safe abstraction for Caffe2
inference). This should be used in ThroughputBenchmark for the purpose
of framework comparison

This specific diff:

There should be no harm in moving transformation code to
OSS. On the advantages side we will be able to compare production
Caffe2 setup with PyTorch in the most fair way via
ThroughputBenchmark. This approach avoid any complicated
transformation regirstries. Building those proper would be significant
engineering effort as well as production risk. In the past we had SEVs
related to transforms being turned off due to various refactors. Given
that we don't plan to build any other significant investments into
transformation logic except existing ones (like TVM and Glow), and
those also relate to open-source technologies, I came up to the
conclusion of moving to OSS the whole thing.

Differential Revision: D16267765

ZolotukhinM reviewed (approved):
Approving since this is all caffe2 files, not touching torch sources. If we decide to use it in pytorch, we'd have to actually port this stuff to use, say, PyTorch JIT IR instead of nomnigraph/nql.

salexspb committed:
BlackBoxPredictor OSS: open-source NQL and custom transforms (#22877)

Summary:
Overal context: open-source BlackBoxPredictor as the entry
point for inference in Caffe2 (thread safe abstraction for Caffe2
inference). This should be used in ThroughputBenchmark for the purpose
of framework comparison

This specific diff:

There should be no harm in moving transformation code to
OSS. On the advantages side we will be able to compare production
Caffe2 setup with PyTorch in the most fair way via
ThroughputBenchmark. This approach avoid any complicated
transformation regirstries. Building those proper would be significant
engineering effort as well as production risk. In the past we had SEVs
related to transforms being turned off due to various refactors. Given
that we don't plan to build any other significant investments into
transformation logic except existing ones (like TVM and Glow), and
those also relate to open-source technologies, I came up to the
conclusion of moving to OSS the whole thing.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/22877

Test Plan:
did a bunch of unit tests locally and now

waitforsandcaslte

AdFinder canary:
https://our.intern.facebook.com/intern/ads/canary/419623727275650390

adindexer:
https://our.intern.facebook.com/intern/ads/canary/419623750891549182

prospector:
https://our.intern.facebook.com/intern/ads/canary/419623766191618404\
seems noisy, launched one more:
https://our.intern.facebook.com/intern/ads/canary/419644899887610977

Differential Revision: D16267765

Pulled By: salexspb

fbshipit-source-id: 25a1cc8d973e1fefc08583dd62d39d0e2ac45a75
